docs/way-of-working/guides/testing/testing.md

```diff
@@ -1 +1,132 @@
+
+- Rank these reasons in terms of importance in our context according to you
+- Any other reasons we should write tests?
+
+:::
+
+
+### The ideal scenario
+
+(Mozafar): we have a **high coverage** of tests, _integration-like_ with the most crucial scenarios tested in a thin layer of _end to end_ tests. Regressions and breaking changes in the API are caught early by tests, and refactoring made easy.
+
+:::warning[Discussion]
+
+- How does the ideal scenario look for you? some other possible prompts: TDD, cucumber, contract testing, end to end, pair programming... 
+
+:::
+
+:::warning[Discussion]
+
+- What should be a high coverage in our context?
+- What is unit vs integration tests in the frontend?
+- How to avoid breaking changes in the API?
+    - Contract testing?
+- Should the ideal scenario depend on the project type: UI library vs Libraries vs apps?
+- How we do end to end tests?
+:::
+
+### Why do we not write tests?
+When we don't write tests,
+
+- Time pressure
+- Not very confident or comfortable with writing tests..
+- Cost of setting up tests when the project has none.
+- Other reasons?
+
+:::warning[Discussion]
+
+- What are other reasons?
+- What can actions can we take to address these?
+:::
+
+
+## Current anti-patterns
+
+- **Inversed test pyramid**: we often over-rely on end to end testing, 
+
+- **Brittle tests**: Tests with too many mocked dependencies that are hard to maintain and don't test integration between components.
+    - **Snapshots** (discussion point): Snapshots are particularly brittle and close to implementation details.
+
+- **Mocked responses in end to end tests**
+
+- **No tests at all!**: For many 
+
+:::warning[Discussion]
+
+- What other anti-patterns?
+
+:::
+
+
+## Common patterns and best practices
+
+This section is to document and point to common patterns and 
+
+:::danger[ToDo]
+add links and context for these patterns, and more concrete examples
+:::
+
+- End to end tests without mocked backend: Scheduler as an example
+- Testing UI components with React Context: (todo examples) 
+- Testing UI components with Zustand 
+- Avoiding flakiness in time and date-related
+
+### react-testing-library patterns
+https://testing-library.com/docs/react-testing-library/intro/
+
+### cypress patterns
+https://docs.cypress.io/guides/references/best-practices
+
+## Other aspects of automated tests
+- Reducing flakiness
+- Tests code quality
+- Using Cucumber or other DSLs
+- Running in CI vs local
+- ....
+
+
+## Readings
+
+- [Five Factor Testing](https://madeintandem.com/blog/five-factor-testing/)
+- [Write tests. Not too many. Mostly integration](https://kentcdodds.com/blog/write-tests)
+- [Practical test pyramid](https://martinfowler.com/articles/practical-test-pyramid.html)
+
+## Proposed actions
+
+:::danger[Proposed actions]
+
+- ? Track code coverage to have a more precise idea of where we are
+
+- ? Rethink e2e tests and move to docker-based solution
+
+- ? Always add a test for bug fixes
+
+- ...
```

Comment:
I think the docs in general are good and cover high level aspects of testing.
One aspect I think is still missing is guidelines on how to write tests, which includes how we want to use tests and how a test should look like, here's an example:

```suggestion
## How to write tests

### Tests as debugging tools

When refactoring or debugging, tests can help us identifying what broke (during refactoring) or analyse and debug the process during which a bug happens (which passed tests and made it into production).

A common pitfall is that we approach tests similarly to how we approach writing production code:

> ## Test code is not like other code
>
> Production applications typically involve thousands to millions of lines of code. They’re too large for humans to  conceptualize all at once. To manage the complexity, language designers have provided mechanisms like functions and class hierarchies that allow developers to think in abstractions.
> 
> Good production code achieves encapsulation. It allows the reader to navigate large systems with ease, diving down into the details or rising to a higher level of abstraction, as needed.
>
> Test code is a different beast. **A good unit test is often small enough that a developer can conceptualize all the logic at once. Adding layers of abstraction to test code increases its complexity. Tests are a diagnostic tool, so they should be as simple and obvious as possible.**
>
> ([Why Good Developers Write Bad Unit Tests - Test code is not like other code](https://mtlynch.io/good-developers-bad-tests/#test-code-is-not-like-other-code); text highlights made by us)

This means that good tests violate a lot of good practices we use and rely on for production code.
DRY test code makes it harder to simply read a test case from top to bottom and understand everything that's happening without having to jump around (either within the file or even across files):

> If a test breaks, the reader should be able to diagnose the problem by reading the test function in a straight line from top to bottom. If they have to jump out of the test to read ancillary code, the test has not done its job.
> [...]
> Before blindly applying DRY to your tests, think about what will make the problem obvious when a test fails. Refactoring may reduce duplication, but it also increases complexity and potentially obscures information when things break.

@TODO: We could quote more suggestions from the linked article ("Go crazy with long test names" & "Embrace magic numbers")

```